optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/manager/DefaultSolverManager.java

```diff
@@ -0,0 +1,188 @@
+        } else {
+            solverFactory = SolverFactory.createFromXmlResource(solverConfigResource);
+        }
+
+        problemIdToSolverTaskMap = new ConcurrentHashMap<>();
+        int numAvailableProcessors = Runtime.getRuntime().availableProcessors();
+        int activeThreadCount = Math.max(numAvailableProcessors - 2, 1);
+        logger.info("Number of available processors: {}. Active thread count: {}.", numAvailableProcessors, activeThreadCount);
+
+        if (threadFactory != null) {
+            solverFactory.getSolverConfig().setThreadFactoryClass(threadFactory.getClass());
+            solverExecutorService = Executors.newFixedThreadPool(activeThreadCount, threadFactory); // #threads < #cpus implies a FIFO scheduling policy
+            eventHandlerExecutorService = Executors.newSingleThreadExecutor(threadFactory);
+        } else {
+            solverExecutorService = Executors.newFixedThreadPool(numAvailableProcessors - 1);
+            eventHandlerExecutorService = Executors.newSingleThreadExecutor();
+        }
+    }
+
+    @Override
+    public void solve(Object problemId,
+                      Solution_ planningProblem,
+                      Consumer<Solution_> onBestSolutionChangedEvent,
+                      Consumer<Solution_> onSolvingEnded) {
+        solve(problemId, planningProblem, onBestSolutionChangedEvent, onSolvingEnded, null);
+    }
+
+    @Override
+    public void solve(Object problemId,
+                      Solution_ planningProblem,
+                      Consumer<Solution_> onBestSolutionChangedEvent,
+                      Consumer<Solution_> onSolvingEnded,
+                      Consumer<Throwable> onException) {
+        SolverTask<Solution_> newSolverTask;
+        synchronized (this) {
+            if (isProblemSubmitted(problemId)) {
+                throw new IllegalArgumentException("Problem (" + problemId + ") already exists.");
+            }
+            newSolverTask = new SolverTask<>(problemId, solverFactory.buildSolver(), planningProblem);
+            problemIdToSolverTaskMap.put(problemId, newSolverTask);
+            logger.info("A new solver task was created with problemId ({}).", problemId);
+        }
+
+        // TODO implement throttling
+        if (onBestSolutionChangedEvent != null) {
+            newSolverTask.addEventListener(
+                    bestSolutionChangedEvent ->
+                            eventHandlerExecutorService.submit(
+                                    () -> onBestSolutionChangedEvent.accept(bestSolutionChangedEvent.getNewBestSolution())));
+        }
+
+        CompletableFuture<Solution_> solverFuture = CompletableFuture.supplyAsync(newSolverTask::startSolving, solverExecutorService);
+        solverFuture.handle((solution_, throwable) -> {
+            if (throwable != null) {
+                logger.error("Exception while solving problem (" + problemId + ").", throwable.getCause());
```

Comment:
String concatenation in logging statements should be replaced with `{}` placeholders .